File: main/java/RecintosZoo.java
package com.anchietaalbano;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

public class RecintosZoo {

    private final List<Recinto> recintos;
    private final List<Animal> animais;

    public RecintosZoo() {
        // Inicializa os recintos
        recintos = new ArrayList<>();
        recintos.add(new Recinto("Recinto 1", "savana", 5, 10));
        recintos.add(new Recinto("Recinto 2", "floresta", 3, 5));
        recintos.add(new Recinto("Recinto 3", "savana", 2, 7));
        recintos.add(new Recinto("Recinto 4", "rio", 5, 8));

        // Inicializa os animais
        animais = new ArrayList<>();
        animais.add(new Animal("LEAO", 3, List.of("savana")));
        animais.add(new Animal("LEOPARDO", 2, List.of("savana")));
        animais.add(new Animal("CROCODILO", 3, List.of("rio")));
        animais.add(new Animal("MACACO", 1, List.of("savana", "floresta")));
        animais.add(new Animal("GAZELA", 2, List.of("savana")));
        animais.add(new Animal("HIPOPOTAMO", 4, List.of("savana", "rio")));
    }

    public ResultadoAnalise analisaRecintos(String tpanimal, int quantidade) {
        // Inicializa o objeto de resultado
        ResultadoAnalise resultado = new ResultadoAnalise(null, new ArrayList<>());

        // Seleciona o animal com base na espécie
        Optional<Animal> selecionarAnimal = animais.stream()
                .filter(animal -> animal.getEspecie().equalsIgnoreCase(tpanimal))
                .findFirst();

        // Se o animal não for encontrado, retorna um erro
        if (selecionarAnimal.isEmpty()) {
            return new ResultadoAnalise("Animal inválido", null);
        }

        Animal animal = selecionarAnimal.get();

        // Valida a quantidade de animais
        if (quantidade <= 0) {
            return new ResultadoAnalise("Quantidade inválida", null);
        }

        // Filtra os recintos que têm espaço livre suficiente e correspondem ao bioma do animal
        List<Recinto> recintosFiltrados = recintos.stream()
                .filter(recinto -> recinto.getEspacoLivre() >= quantidade &&
                        animal.getBioma().contains(recinto.getBioma()))
                .toList();

        // Se não houver recintos viáveis, retorna um erro
        if (recintosFiltrados.isEmpty()) {
            return new ResultadoAnalise("Não há recinto viável", null);
        }

        // Mapeia os recintos viáveis para strings formatadas
        List<String> recintosViaveis = recintosFiltrados.stream()
                .map(recinto -> String.format("%s (bioma: %s, espaço livre: %d, total: %d)",
                        recinto.getNome(), recinto.getBioma(), recinto.getEspacoLivre(), recinto.getCapacidadeTotal()))
                .collect(Collectors.toList());

        // Retorna o resultado com os recintos viáveis
        return new ResultadoAnalise(null, recintosViaveis);
    }


    public static void main(String[] args) {
        // Instancia o RecintosZoo
        RecintosZoo zoo = new RecintosZoo();

        // Exemplo de análise de recintos
        ResultadoAnalise resultado = zoo.analisaRecintos("LEAO", 2);

        // Exibe o resultado
        if (resultado.getErro() != null) {
            System.out.println("Erro: " + resultado.getErro());
        } else {
            System.out.println("Recintos viáveis:");
            resultado.getRecintosViaveis().forEach(System.out::println);
        }
    }
}

File: src/recintos-zoo.js
class Animal {
    constructor(especie, tamanho, biomas) {
        this.especie = especie;
        this.tamanho = tamanho;
        this.biomas = biomas;  
    }
}

class RecintosZoo {

    constructor() {
        this.recintos = [
            { nome: 'Recinto 1', bioma: ['savana'], espacoLivre: 7, capacidadeTotal: 10, animais: ['macaco', 'macaco', 'macaco'] },
            { nome: 'Recinto 2', bioma: ['floresta'], espacoLivre: 5, capacidadeTotal: 5, animais: [] },
            { nome: 'Recinto 3', bioma: ['savana', 'rio'], espacoLivre: 5, capacidadeTotal: 7, animais: ['gazela'] },
            { nome: 'Recinto 4', bioma: ['rio'], espacoLivre: 8, capacidadeTotal: 8, animais: [] },
            { nome: 'Recinto 5', bioma: ['savana'], espacoLivre: 6, capacidadeTotal: 9, animais: ['leao'] }
        ];

        this.animais = [
            new Animal('LEAO', 3, ['savana']),
            new Animal('LEOPARDO', 2, ['savana']),
            new Animal('CROCODILO', 3, ['rio']),
            new Animal('MACACO', 1, ['savana', 'floresta']),
            new Animal('GAZELA', 2, ['savana']),
            new Animal('HIPOPOTAMO', 4, ['savana', 'rio'])
        ];
    }

    analisaRecintos(tpanimal, quantidade) {
        let resultado = { erro: null, recintosViaveis: [] };

        const selecionarAnimal = this.animais.find(animal => animal.especie === tpanimal);

        if (!selecionarAnimal) {
            return { erro: "Animal inválido" };
        }

        if (quantidade <= 0) {
            return { erro: "Quantidade inválida" };
        }

        let espacoNecessario = selecionarAnimal.tamanho * quantidade;

    
        let recintosFiltrados = this.recintos.filter(recinto => {
            //let espacoExtra = (recinto.animais.length > 0 && recinto.animais.every(animal => animal === selecionarAnimal.especie)) ? 1 : 0;
            let espacoDisponivel = recinto.espacoLivre;

            if (espacoDisponivel < espacoNecessario) return false;
            if (!selecionarAnimal.biomas.some(bioma => recinto.bioma.includes(bioma))) return false;
            
            if (selecionarAnimal.especie === 'HIPOPOTAMO' && recinto.bioma.length < 2) return false;
            if (selecionarAnimal.especie === 'MACACO' && recinto.animais.length === 0) {
                return true; 
            }
            if (['LEAO', 'LEOPARDO', 'CROCODILO'].includes(selecionarAnimal.especie) && recinto.animais.length > 0) {
                return recinto.animais.every(animal => animal === selecionarAnimal.especie);
            }

            return true;
        });

        if (recintosFiltrados.length === 0) {
            resultado.erro = "Não há recinto viável";
            resultado.recintosViaveis = null;
        }else {
            resultado.recintosViaveis = recintosFiltrados.map(recinto => {
                let espacoLivreAposAdicionar = recinto.espacoLivre - espacoNecessario;
            
                // Verifica se já há outras espécies no recinto
                const especiesDiferentes = new Set(recinto.animais);
                especiesDiferentes.add(selecionarAnimal.especie); // Adiciona a espécie do novo animal

                // Se houver mais de uma espécie no recinto, adiciona 1 ao espaço necessário
                if (especiesDiferentes.size > 1) {
                    espacoLivreAposAdicionar -= 1;
                }
                    //if (recinto.animais.length > 0 && recinto.animais.some(animal => animal.especie === selecionarAnimal.especie)) {
                    //    espacoLivreAposAdicionar -= espacoNecessario; // Ajusta o espaço extra caso seja necessário
                    //}    
                
                return `${recinto.nome} (espaço livre: ${espacoLivreAposAdicionar} total: ${recinto.capacidadeTotal})`;
            });
        }
        return resultado;
    }
}

export { RecintosZoo as RecintosZoo };
